File: csyd_102/src/ssfilter.cpp
// SSFilter.cp- basic FIR filter

#include "ss.h"
#include "ssfilter.h"

SSFilter::SSFilter(ModList * mList, short h, short v) : SSModule(MT_Filter, mList, h, v)
{
	InitExp(&a0Exp, "0.5");
	InitExp(&a1Exp, "0.5");
	InitExp(&a2Exp, "0");
	InitExp(&b1Exp, "0");
	InitExp(&b2Exp, "0");

	flags = 0;
	DescribeLink(0, "Signal to Filter", "sig",0,0,0xFFFF);
	DescribeLink(1, "Control Signal", "ctl",	0xFFFF,0x1111,0xeeee);
	DescribeLink(2, "Alt Control Sig #1", "ctl1",	0xFFFF,0x2222,0xdddd);
	DescribeLink(3, "Alt Control Sig #2", "ctl2",	0xFFFF,0x3333,0xcccc);
	DescribeLink(4, "Alt Control Sig #3", "ctl3",	0xFFFF,0x4444,0xbbbb);
	DescribeLink(5, "Alt Control Sig #4", "ctl4",	0xFFFF,0x5555,0xaaaa);
	DescribeLink(6, "Alt Control Sig #5", "ctl5",	0xFFFF,0x6666,0x9999);
	DescribeLink(7, "Alt Control Sig #6", "ctl6",	0xFFFF,0x7777,0x8888);
	DescribeLink(8, "Alt Control Sig #7", "ctl7",	0xFFFF,0x8888,0x7777);
	DescribeLink(9, "Alt Control Sig #8", "ctl8",	0xFFFF,0x9999,0x6666);
}

void	SSFilter::Copy(SSModule *ss)
{
	SSFilter *sa = (SSFilter *) ss;
	SSModule::Copy(ss);
	CopyExp(&sa->a0Exp, &a0Exp);
	CopyExp(&sa->a1Exp, &a1Exp);
	CopyExp(&sa->a2Exp, &a2Exp);
	CopyExp(&sa->b1Exp, &b1Exp);
	CopyExp(&sa->b2Exp, &b2Exp);
	this->flags = sa->flags;
}

/*
void SSFilter::Save(FILE* ar)
{
	SSModule::Save(ar);
	WriteFileLine(ar,"FLT %d\r",flags);
  WriteFileLine(ar,"FLTA0 %s\r",a0Exp.exp);
  WriteFileLine(ar,"FLTA1 %s\r",a1Exp.exp);
  WriteFileLine(ar,"FLTA2 %s\r",a2Exp.exp);
  WriteFileLine(ar,"FLTB1 %s\r",b1Exp.exp);
  WriteFileLine(ar,"FLTB2 %s\r",b2Exp.exp);
}
*/

void SSFilter::Load(FILE* ar) 
{
	char				*p,tbuf[256];
	p = parList->GetNextInputLine(ar,"FLT",tbuf);
	sscanf(p, "FLT %d",&flags);
	LoadExp(ar,"FLTA0",&a0Exp);
	LoadExp(ar,"FLTA1",&a1Exp);
	LoadExp(ar,"FLTA2",&a2Exp);
	LoadExp(ar,"FLTB1",&b1Exp);
	LoadExp(ar,"FLTB2",&b2Exp);
}

ssfloat SSFilter::GenerateOutput(SSModule *callingMod)
{
	ssfloat	x,y,a0,a1,a2,b1,b2;

	a0 = SolveExp(&a0Exp,callingMod);
	a1 = SolveExp(&a1Exp,callingMod);
	a2 = SolveExp(&a2Exp,callingMod);
	b1 = SolveExp(&b1Exp,callingMod);
	b2 = SolveExp(&b2Exp,callingMod);

	x = MixInputs(0, callingMod);

	y = a0*x + a1*oldX[0] + a2*oldX[1] +
				b1*oldY[0] + b2*oldY[1];

	oldX[1] = oldX[0];
	oldX[0] = x;
	
	oldY[1] = oldY[0];
	oldY[0] = y;
  lastRightSample = y;
	return y;
}

void SSFilter::Reset(SSModule *callingMod)
{
	SSModule::Reset(callingMod);
	ResetExp(&a0Exp, callingMod);
	ResetExp(&a1Exp, callingMod);
	ResetExp(&a2Exp, callingMod);
	ResetExp(&b1Exp, callingMod);
	ResetExp(&b2Exp, callingMod);
	oldX[0] = 0;
	oldX[1] = 0;
	oldY[0] = 0;
	oldY[1] = 0;
}

File: jsyd_102/src/SSFilter.java
import java.io.*;
import java.util.*;

public class SSFilter extends SSModule implements SydConstants
{
  static final int FF_Recursive = 1;

  public ExpRec    a0Exp = InitExp("0.5"),
            a1Exp = InitExp("0.5"),
            a2Exp = InitExp("0"),
            b1Exp = InitExp("0"),
            b2Exp = InitExp("0");
  public int       flags = 0;

  // dynamic
  double[] oldY = new double[2],
           oldX = new double[2];

  public SSFilter(ModList mList)
  {
    super(MT_Filter, mList);
    DescribeLink(0, "Signal to Filter", "sig");
    DescribeLink(1, "Control Signal", "ctl");
    DescribeLink(2, "Alt Control Sig #1", "ctl1");
    DescribeLink(3, "Alt Control Sig #2", "ctl2");
    DescribeLink(4, "Alt Control Sig #3", "ctl3");
    DescribeLink(5, "Alt Control Sig #4", "ctl4");
    DescribeLink(6, "Alt Control Sig #5", "ctl5");
    DescribeLink(7, "Alt Control Sig #6", "ctl6");
    DescribeLink(8, "Alt Control Sig #7", "ctl7");
    DescribeLink(9, "Alt Control Sig #8", "ctl8");
 }

  void  Copy(SSModule ss)
  {
    SSFilter sa = (SSFilter ) ss;
    super.Copy(ss);
    CopyExp(sa.a0Exp, a0Exp);
    CopyExp(sa.a1Exp, a1Exp);
    CopyExp(sa.a2Exp, a2Exp);
    CopyExp(sa.b1Exp, b1Exp);
    CopyExp(sa.b2Exp, b2Exp);
    this.flags = sa.flags;
  }


  void Save(PrintWriter ar) throws IOException
  {
    super.Save(ar);
    ar.println("FLT " + flags);
    ar.println("FLTA0 " + a0Exp.exp);
    ar.println("FLTA1 " + a1Exp.exp);
    ar.println("FLTA2 " + a2Exp.exp);
    ar.println("FLTB1 " + b1Exp.exp);
    ar.println("FLTB2 " + b2Exp.exp);
  }

  void Load(BufferedReader ar) throws IOException
  {
    flags = Integer.parseInt(getFirstToken(ar,"FLT"));
    a0Exp = LoadExp(ar,"FLTA0");
    a1Exp = LoadExp(ar,"FLTA1");
    a2Exp = LoadExp(ar,"FLTA2");
    b1Exp = LoadExp(ar,"FLTB1");
    b2Exp = LoadExp(ar,"FLTB2");
  }

  double GenerateOutput(SSModule callingMod)
  {

    double a0 = SolveExp(a0Exp,callingMod);
    double a1 = SolveExp(a1Exp,callingMod);
    double a2 = SolveExp(a2Exp,callingMod);
    double b1 = SolveExp(b1Exp,callingMod);
    double b2 = SolveExp(b2Exp,callingMod);

    double x = MixInputs(0, callingMod);

    double y = a0*x + a1*oldX[0] + a2*oldX[1] +
          b1*oldY[0] + b2*oldY[1];

    oldX[1] = oldX[0];
    oldX[0] = x;

    oldY[1] = oldY[0];
    oldY[0] = y;
    lastRightSample = y;
    return y;
  }

  void Reset(SSModule callingMod)
  {
    super.Reset(callingMod);
    ResetExp(a0Exp, callingMod);
    ResetExp(a1Exp, callingMod);
    ResetExp(a2Exp, callingMod);
    ResetExp(b1Exp, callingMod);
    ResetExp(b2Exp, callingMod);
    oldX[0] = 0;
    oldX[1] = 0;
    oldY[0] = 0;
    oldY[1] = 0;
  }

}